Fix memory fault during scaling of singular matrix

chrhansk opened the pull request:
Should fix #200

jfowkes commented:
@chrhansk unfortunately this change seems to break the main SSIDS test on all platforms.

mjacobse commented:
Function `hungarian_match` is not only called from the scaling API, but also for the matching-based METIS ordering of SSIDS. Function `mo_match` in `match_order.f90` expects unmatched entries to be signaled by negative values: https://github.com/ralna/spral/blob/077bd63ad308cca5be8a5e1f937457ebd6097000/src/match_order.f90#L571

So I would suggest not to change the behaviour of `hungarian_match`, but instead fix how the unsymmetric scaling code deals with what `hungarian_match` returns.

Alternatively one could tackle https://github.com/ralna/spral/blob/077bd63ad308cca5be8a5e1f937457ebd6097000/src/match_order.f90#L10-L13 but that might require quite a bit of refactoring?

chrhansk commented:
I was not aware of the problem. I tries to zero out the problematic entries manually in the postprocessing function. There still seems to be a problem in `ssmfe_ciface_test` though. Problem is that I cannot reproduce the error on my system (tests pass without any memory issues). Do you have any ideas?

jfowkes commented:
Many thanks @chrhansk, the intermittent SSMFE C test failure is #204 (nothing to do with your changes) which annoyingly I also cannot reproduce on my system making it very difficult to debug and fix. @mjacobse could you review?

mjacobse reviewed (changes requested):
Thanks, this is certainly an improvement over the current behavior that causes a segfault. However I am not sure the solution of doing the negative to zero conversion in the postprocessing function `match_postproc` is ideal:

* It gives unnecessary responsibility to `match_postproc` and requires changing the `match` argument to `intent(inout)` which to my mind makes the calling contract less clear
* When using the auction method, the conversion in `match_postproc` is done redundantly for a second time after https://github.com/ralna/spral/blob/662c7ac82ad89265b4854fbac8a26b2e2ffb4e38/src/scaling.f90#L1487-L1488 already did it

Instead, I think it would be better to do the conversion in `hungarian_wrapper` before calling `match_postproc`. This is pretty minor and invisible to users though, so perhaps not relevant enough on its own.

More relevant though is that the way unmatched entries are returned from `hungarian_scale_sym` and `hungarian_scale_unsym` is now inconsistent. The unsymmetric version will now return 0 while the symmetric one continues to return negative entries, since the singular symmetric case does not call `match_postproc` as seen here: https://github.com/ralna/spral/blob/662c7ac82ad89265b4854fbac8a26b2e2ffb4e38/src/scaling.f90#L679-L686 Curiously, the current documentation incorrectly claims to return zero for unmatched in both cases ([`hungarian_scale_sym`](https://github.com/ralna/spral/blob/077bd63ad308cca5be8a5e1f937457ebd6097000/docs/Fortran/scaling.rst?plain=1#L401-L402) and [`hungarian_scale_unsym`](https://github.com/ralna/spral/blob/077bd63ad308cca5be8a5e1f937457ebd6097000/docs/Fortran/scaling.rst?plain=1#L420-L421)), perhaps copy-pasted from the description of the auction method for which it is correct. Options I can think of:

* Do the conversion from negative to zero on a temporary copy of `match`. That way, both `hungarian_scale_sym` and `hungarian_scale_unsym` continue to return negative entries. With this option, the incorrect documentation for both cases should be fixed (perhaps in a separate issue). 
* Accept this inconsistency. With this option, the incorrect documentation for the symmetric case should be fixed (perhaps in a separate issue)
* Make the symmetric case work with and return zeros for unmatched entries too. The necessary changes should be limited to `hungarian_wrapper` and should work well with the changes for the unsymmetric case (when done in `hungarian_wrapper`, which would add a major reason for doing so to the above). Because of that it might make sense to change both at once instead of in a seperate issue.

The latter two options would break potential users who are relying on the negative entries (despite the wrong documentation) or would like to do so in the future, but it would introduce consistency with how the auction methods returns the matching and with the documented behavior. Not sure what's the best call here.

chrhansk commented on tests/scaling.f90:
```diff
@@ -573,6 +574,55 @@ end subroutine test_equilib_unsym_random
+
+  allocate(a%ptr(n+1))
+  allocate(a%row(nz), a%val(nz))
+  allocate(rscaling(m), cscaling(n), match(n))
```
I am not sure about that. The [docs](https://ralna.github.io/spral/_build/html/Fortran/scaling.html#f/spral_scaling/hungarian_scale_unsym) state that it should be `n` (similarly for all matching algorithms). I am not exactly sure why to be honest though, but calling it with `m` in the unit test causes a segfault on my machine.

mjacobse replied:
Hm you are right, I do get invalid write with valgrind for `match(m)`. But when doing `match(n)`, the last two entries in the example of that test are left uninitialized which surely is not intended either? The signature for `hungarian_scale_unsym` does use `m`: https://github.com/ralna/spral/blob/077bd63ad308cca5be8a5e1f937457ebd6097000/src/scaling.f90#L188

Something seems quite wrong here... :(

mjacobse replied:
See #207

mjacobse replied:
Turns out that `match(m)` is correct but that this happened to reveal a secondary bug which should probably be fixed before applying the changes proposed in this PR, see https://github.com/ralna/spral/issues/200#issuecomment-2176922979.

mjacobse commented on tests/scaling.f90:
```diff
@@ -573,6 +574,55 @@ end subroutine test_equilib_unsym_random
+  a%n = n
+  a%m = m
+
+  a%ptr(1:n+1)          = (/   1,   3,   5,   5,   5,   7 /)
```
Should this be `1, 3, 5, 6, 6, 7` as in #200? As it stands, this matrix would have a duplicate entry.

chrhansk commented:
Looking back at this: In particular in regards to #205: What is the proposed solution that you are converging on? As I understand https://github.com/ralna/spral/issues/200#issuecomment-2178124090 The lines setting the unmatched rows to negative values should be commented out, which to me implies that the values of unmatched rows should be set to zero instead. Am I correct in this regard?

Such a change would necessitate corresponding changes in the implementation of SSIDS, as mentioned here: https://github.com/ralna/spral/pull/205#issuecomment-2169174578

Should I make those changes to the scaling and within SSIDS?

chrhansk commented:
I adjusted the scaling accordingly and added a test for the symmetric singular case. There is however still the problematic part [here](https://github.com/ralna/spral/blob/f0a82e6a682b2b98a08ecd65a58b69f10b95a005/src/scaling.f90#L1176-L1193) mentioned in #200. In this case it causes a segfault (if the match array has a size of `m` < `n`), so this needs to be addressed.

jfowkes commented:
Many thanks @chrhansk, the suggestion has been to comment this problematic section out as it should no longer be required if we now return zero for unmatched entries. I guess it's a case of trying that and seeing if anything breaks?

mjacobse commented:
Basically the suggestion was this https://github.com/mjacobse/spral/commit/b815fac89a4ca6bcf5ac8a5801cb4efb1a087fad and indeed, it seems to be all that's needed to fix all issues at once. Personally I would like to see randomly generated singular tests to confirm, since all the random tests right now are nonsingular.

jfowkes commented:
Indeed many thanks! @chrhansk I suggest we apply https://github.com/mjacobse/spral/commit/b815fac89a4ca6bcf5ac8a5801cb4efb1a087fad to this PR and add a randomly generated singular matrix test to verify that things don't break. We think this should be all that is required now.

mjacobse commented on tests/scaling.f90:
```diff
@@ -597,6 +599,104 @@ end subroutine test_equilib_unsym_random
+  endif
+
+  if(match(ising) .ne. 0) then
+     write(*, "(a, i5, a, i5)") "Singular column ", ising, "matched to ", match(ising)
```
It's correct, but maybe "Singular row" would be more clear since match has size m?

jfowkes replied:
Agreed that's clearer.

mjacobse commented on tests/scaling.f90:
```diff
@@ -597,6 +599,104 @@ end subroutine test_equilib_unsym_random
+subroutine test_hungarian_sym_singular
+  integer :: m = 3
+  integer :: n = 3
+  integer :: nz = 6
```
nz = 2?

mjacobse reviewed (approved):
Yeah sure, let's merge it. I had tried to allow singular cases within the functions that do the random tests for scaling, but it turned out to not be so easy so I kinda lost track of it, sorry about that. Perhaps at some point we can look into it again, but for now let's go with this, should be a useful fix.

jfowkes commented:
No worries, generating random singular matrices is not so easy. I'll add your suggestions and merge.

jfowkes commented on src/match_order.f90:
```diff
@@ -611,7 +611,7 @@ subroutine mo_match(n,row2,ptr2,val2,scale,flag,stat,perm)
     do i = 1, n
        j = old_to_new(i)
-       if (j .lt. 0) then
+       if (j .eq. 0) then
```
Indeed, reverted by #231
